pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,567 @@
+// and services that provide the actual backends.
+type Proxier struct {
+	mu         sync.Mutex // protects serviceMap
+	serviceMap map[proxy.ServicePortName]*serviceInfo
+	iptables   utiliptables.Interface
+	hostIP     net.IP
+}
+
+// Proxier implements ProxyProvider
+var _ proxy.ProxyProvider = &Proxier{}
+
+// ErrProxyOnLocalhost is returned by NewProxier if the user requests a proxier on
+// the loopback address. May be checked for by callers of NewProxier to know whether
+// the caller provided invalid input.
+var ErrProxyOnLocalhost = fmt.Errorf("cannot proxy on localhost")
+
+// NewProxier returns a new Proxier given an address on which to listen.
+// Because of the iptables logic, It is assumed that there is only a single
+// Proxier active on a machine.
+// An error will be returned if the proxier cannot be started due to an invalid
+// ListenIP (loopback) or if iptables fails to update or acquire the initial lock.
+// Once a proxier is created, it will keep iptables up to date in the background and
+// will not terminate if a particular iptables call fails.
+func NewProxier(ipt utiliptables.Interface) (*Proxier, error) {
+	hostIP, err := util.ChooseHostInterface()
+	if err != nil {
+		return nil, fmt.Errorf("failed to select a host interface: %v", err)
+	}
+
+	glog.V(2).Infof("Setting proxy IP to %v and initializing iptables", hostIP)
+	return &Proxier{
+		serviceMap: make(map[proxy.ServicePortName]*serviceInfo),
+		iptables:   ipt,
+		hostIP:     hostIP,
+	}, nil
+}
+
+func (proxier *Proxier) getServiceInfo(service proxy.ServicePortName) (*serviceInfo, bool) {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+	info, ok := proxier.serviceMap[service]
+	return info, ok
+}
+
+func (proxier *Proxier) setServiceInfo(service proxy.ServicePortName, info *serviceInfo) {
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+	proxier.serviceMap[service] = info
+}
+
+func (proxier *Proxier) sameConfig(info *serviceInfo, service *api.Service, port *api.ServicePort) bool {
+	if info.protocol != port.Protocol || info.portal.port != port.Port {
+		return false
+	}
+	if !info.portal.ip.Equal(net.ParseIP(service.Spec.ClusterIP)) {
+		return false
+	}
+	if !ipsEqual(info.deprecatedPublicIPs, service.Spec.DeprecatedPublicIPs) {
+		return false
+	}
+	if !api.LoadBalancerStatusEqual(&info.loadBalancerStatus, &service.Status.LoadBalancer) {
+		return false
+	}
+	if info.sessionAffinityType != service.Spec.SessionAffinity {
+		return false
+	}
+	return true
+}
+
+func ipsEqual(lhs, rhs []string) bool {
+	if len(lhs) != len(rhs) {
+		return false
+	}
+	for i := range lhs {
+		if lhs[i] != rhs[i] {
+			return false
+		}
+	}
+	return true
+}
+
+// TODO: what should this be?
+const syncIntervalIptables = 5 * time.Second
+
+// SyncLoop runs periodic work.  This is expected to run as a goroutine or as the main loop of the app.  It does not return.
+func (proxier *Proxier) SyncLoop() {
+	//TODO: should the sync interval differ from Proxier's ?
```

Comment:
This referred to the interval of the userspace proxier sync loop from when they had different names.